src/types/redis_bloom_chain.cc

```diff
@@ -0,0 +1,201 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#include "redis_bloom_chain.h"
+
+namespace redis {
+
+std::string BloomChain::getBFKey(const Slice &ns_key, const BloomChainMetadata &metadata, uint16_t filters_index) {
+  std::string sub_key;
+  sub_key.append(kBloomFilterSeparator);
+  PutFixed16(&sub_key, filters_index);
+
+  std::string bf_key = InternalKey(ns_key, sub_key, metadata.version, storage_->IsSlotIdEncoded()).Encode();
+  return bf_key;
+}
+
+void BloomChain::getBFKeyList(const Slice &ns_key, const BloomChainMetadata &metadata,
+                              std::vector<std::string> *bf_key_list) {
+  bf_key_list->reserve(metadata.n_filters);
+  for (uint16_t i = 0; i < metadata.n_filters; ++i) {
+    std::string bf_key = getBFKey(ns_key, metadata, i);
+    bf_key_list->push_back(std::move(bf_key));
+  }
+}
+
+rocksdb::Status BloomChain::getBloomChainMetadata(const Slice &ns_key, BloomChainMetadata *metadata) {
+  return Database::GetMetadata(kRedisBloomFilter, ns_key, metadata);
+}
+
+rocksdb::Status BloomChain::createBloomChain(const Slice &ns_key, double error_rate, uint32_t capacity,
+                                             uint16_t expansion, BloomChainMetadata *metadata) {
+  metadata->n_filters = 1;
+  metadata->expansion = expansion;
+  metadata->size = 0;
+
+  metadata->error_rate = error_rate;
+  metadata->base_capacity = capacity;
+  metadata->bloom_bytes = BlockSplitBloomFilter::OptimalNumOfBytes(capacity, error_rate);
+
+  BlockSplitBloomFilter block_split_bloom_filter;
+  block_split_bloom_filter.Init(metadata->bloom_bytes);
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisBloomFilter, {"createSBChain"});
```

Comment:
The logic here is related to `WriteBatchExtractor::PutCF`.

```c++
constexpr std::string_view CreateBloomChain = "...";
constexpr std::string_view InsertBloomFilter = "...";
```

And in `src/storage/batch_extractor.cc`, you'd better have the correspond behavior. Like:

```
case kRedisBloomFilter: {
  // ...
}
```

And follow the `kRedisBitmap` style here.

Reply:
Or you can leave a todo in `src/storage/batch_extractor.cc`, I can add an todo in issue tracker

Reply:
Ok, I will fix it in the next patch.

Reply:
You can fix it after all function is done :-) migration might be refactor later